### PR TITLE
loop-util: enable LO_FLAGS_DIRECT_IO by default on loopback devices

### PR DESCRIPTION
**Pull request by poettering**
Fixes: #21003

### CODE REVIEW

File: src/shared/loop-util.c [hunk 1]
```diff
@@ -319,6 +326,14 @@ static int loop_configure(
                               random_u64_range(UINT64_C(240) * USEC_PER_MSEC * n_attempts/64));
         }
 
+        /* LO_FLAGS_DIRECT_IO is a flags we need to configure via explicit ioctls. */
+        if (FLAGS_SET(c->info.lo_flags, LO_FLAGS_DIRECT_IO)) {
+                unsigned long b = 1;
+
+                if (ioctl(fd, LOOP_SET_DIRECT_IO, b) < 0)
+                        log_debug_errno(errno, "Failed to enable direct IO mode on loopback device, ignoring: %m");
```
Comment (bluca) [inline src/shared/loop-util.c]: Do we have a user-readable device/file name here that we can log?

File: src/shared/loop-util.c [hunk 2]
```diff
@@ -444,6 +460,30 @@ int loop_device_make(
...
+                 * that's required. Conversely, if it's off require that O_DIRECT is off too (that's because
+                 * new kernels will implicitly enable LO_FLAGS_DIRECT_IO if O_DIRECT is set).
+                 *
+                 * Our intention here is that LO_FLAGS_DIRECT_IO is the primary knob, and O_DIRECT derived
+                 * from that automatically. */
+
+                direct_io_fd = fd_reopen(fd, (FLAGS_SET(loop_flags, LO_FLAGS_DIRECT_IO) ? O_DIRECT : 0)|O_CLOEXEC|O_NONBLOCK|open_flags);
+                if (direct_io_fd < 0) {
+                        if (!FLAGS_SET(loop_flags, LO_FLAGS_DIRECT_IO))
+                                return log_debug_errno(errno, "Failed to turn off O_DIRECT for file descriptor: %m");
```
Comment (bluca) [inline src/shared/loop-util.c]: Maybe `Failed to reopen file descriptor without O_DIRECT`? "turn off" sounds like an ioctl or suchlike was attempted

File: src/shared/loop-util.c [hunk 3]
```diff
@@ -505,6 +545,28 @@ int loop_device_make(
...
+                /* On older kernels (<= 5.3) it was necessary to set the block size of the loopback block
+                 * device to the logical block size of the underlying file system. Since there was no nice
+                 * way to query the value, we are not bothering to do this however. On newer kernels the
+                 * block size is propagated automatically and does not require intervention from us. We'll
+                 * check here if enabling direct IO worked, to make this easily debuggable however.
+                 *
+                 * (Should anyone really care and actually wants direct IO on old kernels: it might be worth
+                 * enabling direct IO with iteratively larger block sizes until it eventually works.) */
+                if (!FLAGS_SET(info.lo_flags, LO_FLAGS_DIRECT_IO))
+                        log_debug("Unable to enable direct IO mode, proceeding in buffered IO mode.");
```
Comment (bluca) [inline src/shared/loop-util.c]: nit: "Unable to enable" is a bit of a mouthful, maybe "Failed to enable" or "Could not enable"

File: src/shared/loop-util.c [hunk 4]
```diff
@@ -547,15 +643,23 @@ int loop_device_make_by_path(
         /* Passing < 0 as open_flags here means we'll try to open the device writable if we can, retrying
          * read-only if we cannot. */
 
-        fd = open(path, O_CLOEXEC|O_NONBLOCK|O_NOCTTY|(open_flags >= 0 ? open_flags : O_RDWR));
+        loop_flags = loop_flags_mangle(loop_flags);
+
+        /* Let's open with O_DIRECT if we can. But not all file systems support that, hence fall back to
+         * non-O_DIRECT mode automatically, if it fails. */
+        fd = open(path, O_CLOEXEC|O_NONBLOCK|O_NOCTTY|(open_flags >= 0 ? open_flags : O_RDWR)|(FLAGS_SET(loop_flags, LO_FLAGS_DIRECT_IO) ? O_DIRECT : 0));
```
Comment (bluca) [inline src/shared/loop-util.c]: this line is a bit hard to digest with all the conditionals, maybe add line breaks and indent?

File: src/shared/loop-util.c [hunk 5]
```diff
@@ -547,15 +643,23 @@ int loop_device_make_by_path(
...
+        if (fd < 0 && FLAGS_SET(loop_flags, LO_FLAGS_DIRECT_IO)) /* If we had O_DIRECT on, and things failed with that, let's immediately try again without */
+                fd = open(path, O_CLOEXEC|O_NONBLOCK|O_NOCTTY|(open_flags >= 0 ? open_flags : O_RDWR));
```
Comment (bluca) [inline src/shared/loop-util.c]: the same two sets of flags - with and without O_DIRECT - are used multiple times, maybe a const variable to store and reuse would make it more readable? And less error prone for future changes

File: src/shared/loop-util.c [hunk 6]
```diff
@@ -326,6 +326,21 @@ static int loop_configure(
                               random_u64_range(UINT64_C(240) * USEC_PER_MSEC * n_attempts/64));
         }
 
+        /* Work around a kernel bug, where the chaning offset/size of the loopback device doesn't correctly
```
Comment (bluca) [inline src/shared/loop-util.c]: "where changing offset/size" ?

File: src/shared/loop-util.c [hunk 7]
```diff
@@ -531,14 +593,48 @@ int loop_device_make(
...
 
+static uint32_t loop_flags_mangle(uint32_t loop_flags) {
+        int r;
+
+        r = getenv_bool("SYSTEMD_LOOP_DIRECT_IO");
+        if (r < 0 && r != -ENXIO)
+                log_debug_errno(r, "Failed to parse $SYSTEMD_LOOP_DIRECT_IO, ignoring: %m");
+
+        SET_FLAG(loop_flags, LO_FLAGS_DIRECT_IO, r != 0); /* Turn on LO_FLAGS_DIRECT_IO by default, unless explicitly configured to off. */
+        return loop_flags;
```
Comment (yuwata) [inline src/shared/loop-util.c:619]: Not a matter, but you can use `return UPDATE_FLAG(...)`.

File: src/shared/loop-util.c [hunk 8]
```diff
@@ -531,14 +593,48 @@ int loop_device_make(
...
+                int fd,
+                int open_flags,
+                uint64_t offset,
+                uint64_t size,
+                uint32_t loop_flags,
+                LoopDevice **ret) {
+
+        assert(fd >= 0);
+        assert(ret);
+        assert(IN_SET(open_flags, O_RDWR, O_RDONLY));
```
Comment (yuwata) [inline src/shared/loop-util.c:632]: These are not necessary in this function. `_internal()` has them.

File: src/shared/loop-util.c [hunk 9]
```diff
@@ -531,14 +593,48 @@ int loop_device_make(
...
+
+        loop_flags = loop_flags_mangle(loop_flags);
+
+        return loop_device_make_internal(
```
Comment (yuwata) [inline src/shared/loop-util.c:636]: Not a matter, but simply `return loop_device_make_internal(..., loop_flags_mangle(loop_flags), ...);` is ok.

File: src/shared/loop-util.c [hunk 10]
```diff
@@ -547,15 +643,28 @@ int loop_device_make_by_path(
...
+        /* Let's open with O_DIRECT if we can. But not all file systems support that, hence fall back to
+         * non-O_DIRECT mode automatically, if it fails. */
+
+        basic_flags = O_CLOEXEC|O_NONBLOCK|O_NOCTTY;
+        direct_flags = FLAGS_SET(loop_flags, LO_FLAGS_DIRECT_IO) ? O_DIRECT : 0;
+        rdwr_flags = open_flags >= 0 ? open_flags : O_RDWR;
+
+        fd = open(path, basic_flags|direct_flags|rdwr_flags);
+        if (fd < 0 && direct_flags != 0) /* If we had O_DIRECT on, and things failed with that, let's immediately try again without */
+                fd = open(path, basic_flags|rdwr_flags);
```
Comment (yuwata) [inline src/shared/loop-util.c:672]: Maybe better to add a debug log that fallback logic will be used.

File: src/shared/loop-util.c [hunk 11]
```diff
@@ -547,15 +643,28 @@ int loop_device_make_by_path(
...
         if (fd < 0) {
                 r = -errno;
 
                 /* Retry read-only? */
                 if (open_flags >= 0 || !(ERRNO_IS_PRIVILEGE(r) || r == -EROFS))
                         return r;
 
-                fd = open(path, O_CLOEXEC|O_NONBLOCK|O_NOCTTY|O_RDONLY);
+                fd = open(path, basic_flags|direct_flags|O_RDONLY);
```
Comment (yuwata) [inline src/shared/loop-util.c:680]: Also here, debug log opening with O_RDONLY?

**Pull request merged**